### PR TITLE
[go_router] [shell_route] Add observers parameter

### PR DESCRIPTION
**Pull request by angjelkom**
This PR adds `observers` parameter to the `ShellRoute` and passes that to the nested Navigator created by that same `ShellRoute`. 

The need for this PR is because it fixes cases where a `Hero` animation doesn't work when used with nested navigation and it requires a `HeroController` to be used as an observer for that nested Navigator.

fixes https://github.com/flutter/flutter/issues/112095

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by grimatoma — COMMENTED** [review]
LGTM

File: packages/go_router/lib/src/builder.dart
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:176]: I think we need to create a new herocontroller here. It doesn't make sense to ask user to pass in controller if it is needed anyway

Comment (angjelkom) [reply packages/go_router/lib/src/builder.dart:176]: @chunhtai I get your point, but not all apps have the need of a HeroController as it's only needed if user actually uses a Hero Widget, besides this in general solves the need for a middleware for the nested Navigator. I don't see a reason why a ShellRoute should not expose observers for the nested Navigator.

Comment (thipokch) [reply packages/go_router/lib/src/builder.dart:176]: @chunhtai  Another use case, would be for `FirebaseAnalyticsObserver`. Currently, it is only observed for routes at root `GoRouter`, by allowing this `NavigatorObserver` would allow us to track navigation in google analytics.

Comment (thipokch) [reply packages/go_router/lib/src/builder.dart:176]: Shall we do that in a new PR and merge this?

Comment (thipokch) [reply packages/go_router/lib/src/builder.dart:176]: @chunhtai Shall we do that in a new PR and merge this?

Comment (chunhtai) [reply packages/go_router/lib/src/builder.dart:176]: I think we should add it in this PR.

Comment (chunhtai) [reply packages/go_router/lib/src/builder.dart:176]: I would like to clarify why we should do this now. If we merge this as is, people will start to insert their own HeroController so they can make hero to work. It will then be a breaking change if we create the herocontroller by default in the later change

**Comment by chunhtai** [comment]
Hi @angjelkom , are you still planning to work on this pr?

**Comment by angjelkom** [comment]
> Hi @angjelkom , are you still planning to work on this pr?

Yes I will add your suggested change over the weeekend.

**Comment by angjelkom** [comment]
@chunhtai I will need a bit more time, as merging the latest commits from the main branch caused my example to have weird side-effects where the Hero animation doesn't work when going to the next page but when going back it works in a buggy-weird way.

So until i solve that I can't commit the change for HeroControllerScope.

**Comment by chunhtai** [comment]
@angjelkom are still planning on coming back to this pr?

**Comment by angjelkom** [comment]
@chunhtai I can't complete the pull request until [#115832](https://github.com/flutter/flutter/issues/115832) its fixed, its blocking me.

I could use a lower commit but still I would need to test it against the latest commit at the end and currently that's not possible because of the above issue.

**Comment by chunhtai** [comment]
@angjelkom I will take a look at the issue and try to resolve it asap

**Comment by ycherniavskyi** [comment]
@angjelkom, the issue you mentioned has been fixed. Could you update your PR for merging? If you are too busy, I could create an analogic PR. I need it solely for `FirebaseAnalyticsObserver`.

File: packages/go_router/lib/src/builder.dart
Comment (ycherniavskyi) [reply packages/go_router/lib/src/builder.dart:176]: @chunhtai, as for creating the default `HeroController`, how could we determine which static factory method to use for `HeroController` creation (`createCupertinoHeroController` or `createMaterialHeroController`) within design agnostic `go_router`?

**Comment by angjelkom** [comment]
@ycherniavskyi the PR is ready waiting for @chunhtai to review it and approve it.

File: packages/go_router/lib/src/route.dart
Comment (ycherniavskyi) [inline packages/go_router/lib/src/route.dart:465]: I have asked @angjelkom exactly about this part.
`CupertinoApp` and `MaterialApp` use the static methods `createCupertinoHeroController` and `createMaterialHeroController` accordingly (where the last one sets the specific function to `createRectTween` property of created `HeroController`) to create `HeroController`.
Does `HeroController` creation here need to respect default app behavior? And if yes, it seems that we need to move `HeroController` creation from `ShellRoute` class to `RouteBuilder` class and implement it the same way as a specific page type build for a `builder` function (here, I mean `_pageBuilderForAppType` variable that specifies in `_cacheAppType` depending from `elem` app type).

File: packages/go_router/lib/src/builder.dart
Comment (ycherniavskyi) [inline packages/go_router/lib/src/builder.dart:200]: Why does only this `_buildNavigator` method call wrapped with `HeroControllerScope`? Shouldn't all `_buildNavigator` method calls be wrapped with `HeroControllerScope`?

Comment (angjelkom) [reply packages/go_router/lib/src/builder.dart:200]: > 

The issue with the Hero Widgets not animating is only with the nested navigator and that is the ShellRoute, I've tested the top Navigator and for that the Hero widgets work fine, so `HeroControllerScope` is not required there.

File: packages/go_router/lib/src/route.dart
Comment (angjelkom) [reply packages/go_router/lib/src/route.dart:465]: @ycherniavskyi you do have a point there, I will check this after work and let you know, but probably we do need respect default app behaviour and use the platform specific HeroController.

**Comment by arsen** [comment]
any chance this will get merged soon?

**Comment by angjelkom** [comment]
@arsen I will get on it tonight to finish it up.

**Comment by angjelkom** [comment]
@chunhtai as @ycherniavskyi suggested we should use the relevant HeroController based on whether MaterialApp or CupertinoApp is used, in order to do that we need to use the RouteBuilder not the ShellRoute in order to access the context.

The issue is because the RouteBuilder gets rebuild while navigating the HeroController which again breaks the animation, the way I see it we have two options:

Option 1:

Create 3 final variables in the RouteBuilder, one using createMaterialHeroController, one using createCupertinoHeroController and a default HeroController.
Then we can use those to pass the correct controller.
This way the controllers won't be rebuild and the animation would work. (tested)

Option 2:

Have a heroController parameter to the ShellRoute so that the user passes one himself.

Let me know what would you recommend, I can do the change tonight.

File: packages/go_router/lib/src/builder.dart
Comment (matecode) [reply packages/go_router/lib/src/builder.dart:200]: Option2 ;)

**Comment by matecode** [comment]
Option2 Vote!

**Comment by chunhtai** [comment]
GoRouter uses isMaterialApp and isCupertinoApp to infer which Page subclass to use, maybe you can also use the same logic to create hero controllers?

**Comment by angjelkom** [comment]
@chunhtai that's not what I meant, of course I would use isMaterialApp and isCupertinoApp, the situation is that creating the hero controller inside the RouteBuilder build method its not possible, because the build method gets called when navigating which breaks the animation.

So there are two solutions one is creating three final variables on the RouteBuilder containing the material, cupertino and default version of HeroController, and then use the isMaterialApp and isCupertinoApp to use the correct final variable controller.

The other solution is to add a heroController parameter to the ShellRoute and let the end user pass the controller himself.

So I'm asking what would be the better approach looking from api perspective?

Add a heroController parameter or solve it internally?

**Comment by ycherniavskyi** [comment]
@angjelkom, as for **Option 1** - as I understand, we need to create `HeroController` for each `Navigator` (`ShellRoute`) 🤔; as for **Option 2** - as I understand @chunhtai already rejects it.

What if we add the optional `HeroController` property in `GoRouterState`, then assign it within `buildState` method of `RouteBuilder` (where we can use `isMaterialApp` and `isCupertinoApp`) and finally use it for for `HeroControllerScope` as you did in MR. If changing `GoRouterState` isn't desirable, then an additional registry or even a simple map could be added to `RouteBuilder` for mapping corresponding `HeroController` with `ShellRoute`.

**Comment by chunhtai** [comment]
Oh I see what's the issue now. can you create a cache in the RouteBuilder to store navkey to HeroController map? that way you don't recreate the herocontroller if it is in the cache.

I am not sure if this is the best solution because I think the real solution is to refactor herocontroller to not depend on Navigator, see https://github.com/flutter/flutter/issues/54200. but if everything is kept private, then I think it will be fine.

It can be smart about updating the cache when isMaterialApp changes, but I don't think that is a common use case.

**Comment by angjelkom** [comment]
@chunhtai I completed the PR, but the dart_unit_tests are failing, locally running the commands work ok, what am I missing here?

File: packages/go_router/lib/src/route.dart
Comment (ycherniavskyi) [inline packages/go_router/lib/src/route.dart:465]: I am not sure that it is the right concept to resolve `HeroController` issue. As for me, `GoRoute` is not guilty of knowing about `HeroController` at all.

@angjelkom Why not "create a cache in the RouteBuilder to store navkey to HeroController map" (as @chunhtai suggested)? It could be a simple `Map<GlobalKey<NavigatorState>, HeroController> _goHeroCache;` property in `RouteBuilder`. And when `HeroController` is needed, we check the cache and, if necessary, create the right one based on `isMaterialApp`/`isCupertinoApp`.

File: packages/go_router/lib/src/misc/go_hero.dart
Comment (ycherniavskyi) [inline packages/go_router/lib/src/misc/go_hero.dart:20]: Prematurely creating a tree instance of `HeroController` seems redundant here.

**Comment by chunhtai** [comment]
looked like the ci is blocked by https://github.com/flutter/packages/pull/2977

**Comment by angjelkom** [comment]
@ycherniavskyi Thanks for the suggestion, @chunhtai I added the cache solution, yes looks like CI is blocked

File: packages/go_router/lib/src/builder.dart
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:55]: At the end of the build method, we need to clean up the cache that doesn't have associated navigatorKey to prevent memory leak

File: packages/go_router/lib/src/builder.dart
Comment (angjelkom) [inline packages/go_router/lib/src/builder.dart:458]: @chunhtai Added the clean up here, as this will get called only if the NavigatorKey isn't assigned to the cache.

Comment (ycherniavskyi) [reply packages/go_router/lib/src/builder.dart:458]: Wow, it seems too aggressive cache-clearing policy 🙂. Here you clear all cache on any new element added.

I could propose the following cache-clearing policy that executes at the end of `buildPages` method:
```dart
  List<Page<Object?>> buildPages(
      BuildContext context,
      RouteMatchList matchList,
      PopPageCallback onPopPage,
      bool routerNeglect,
      GlobalKey<NavigatorState> navigatorKey,
      Map<Page<Object?>, GoRouterState> registry) {
    final Map<GlobalKey<NavigatorState>, List<Page<Object?>>> keyToPage =
        <GlobalKey<NavigatorState>, List<Page<Object?>>>{};
    try {
      _buildRecursive(context, matchList, 0, onPopPage, routerNeglect,
          keyToPage, navigatorKey, registry);
      return keyToPage[navigatorKey]!;
    } on _RouteBuilderError catch (e) {
      return <Page<Object?>>[
        _buildErrorPage(context, e, matchList.uri),
      ];
    } finally {
      _goHeroCache.removeWhere((key, _) => !keyToPage.keys.contains(key));
    }
  }
```

Comment (angjelkom) [reply packages/go_router/lib/src/builder.dart]: Thanks, added your proposed cache-policy.

File: packages/go_router/lib/src/builder.dart
Comment (ycherniavskyi) [reply packages/go_router/lib/src/builder.dart:55]: @chunhtai, it seems that cache clean-up logic was added. What else do we need to improve before marge?

File: packages/go_router/lib/src/pages/cupertino.dart
Comment (chunhtai) [inline packages/go_router/lib/src/pages/cupertino.dart:14]:
```suggestion
/// Creates a Cupertino HeroController.
```

File: packages/go_router/lib/src/pages/material.dart
Comment (chunhtai) [inline packages/go_router/lib/src/pages/material.dart:15]:
```suggestion
/// Creates a Material HeroController.
```

File: packages/go_router/test/builder_test.dart
Comment (chunhtai) [inline packages/go_router/test/builder_test.dart:276]: Can you also add a check that hero will be trigger if under shell route?

**Review by chunhtai — APPROVED** [review]
LGTM, just some nits

**Comment by ycherniavskyi** [comment]
@angjelkom, could you please fix "some nits" mentioned by @chunhtai, so this PR finally could be merged?

File: packages/go_router/test/builder_test.dart
Comment (angjelkom) [reply packages/go_router/test/builder_test.dart:276]: @chunhtai I'm sorry holidays are here at the moment, if you could give me a suggestion how to do it, I will commit that.

**Comment by VusDanylo** [comment]
Will it be merged soon?

**Comment by ycherniavskyi** [comment]
@VusDanylo could you please be more specific regarding the term "soon"? :) When exactly could we expect to get this PR merged?

**Comment by angjelkom** [comment]
@ycherniavskyi nits have been fixed, except for the better unit test, I am on holidays now I can have look during this weekend. If anyone can provide a test snippet now I can merge it sometimes today.

File: packages/go_router/lib/src/builder.dart
Comment (jascodes) [reply packages/go_router/lib/src/builder.dart]: For me, it causes a very significant performance hit. Not entirely sure what is the reason.

**Comment by jonasbark** [comment]
@chunhtai sorry for pinging you but this PR feels like a very important fix. Using shell routes causes any analytic route observers (e.g. firebase or custom ones) to fail working correctly.

This PR has been open for months and imho it gives go_router maintenance a bad look - as well as any outside contribution. I'm also seeing the same issue for other PRs, such as #2650 or go_router_builder support for shell routes.

Is there anything the open source community can do to help?

**Comment by ahmednfwela** [comment]
can't this PR just remove the additional `HeroControllerScope` feature and instead focus on the `observers` one ?

**Comment by ycherniavskyi** [comment]
As I understand, the last barrier before merging is [extending the hero controller test](https://github.com/flutter/packages/pull/2664#discussion_r1061700389). @chunhtai am I correct?

File: packages/go_router/test/builder_test.dart
Comment (chunhtai) [reply packages/go_router/test/builder_test.dart:276]: You can have two sub goRoutes under a shellroute

lets say GoRouteA and GoRouteB, makes sure both GoRoute Build a Hero widget with the same tag and with flightShuttleBuilder to build some thing.

When they you do a `go` from GoRouteA to GoRouteB. During the page transition, you can check whether the widget that built in flightShuttleBuilder can be found through Find.

**Review by chunhtai — COMMENTED** [review]
Sorry I just got back from vacation.

**Comment by chunhtai** [comment]
> As I understand, the last barrier before merging is [extending the hero controller test](https://github.com/flutter/packages/pull/2664#discussion_r1061700389). @chunhtai am I correct?

correct

**Comment by kamami** [comment]
When will this be merged?

File: packages/go_router/test/builder_test.dart
Comment (flodaniel) [reply packages/go_router/test/builder_test.dart:276]: I tried to write a test based on your comment @chunhtai, but it keeps failing. I am unsure if the pumping step works correctly, or if the `router.go` navigation completes before I can check if the key is present.  Currently it keeps failing, because it does not find the heroKey after calling `.go` and `.pump`

 ```
testWidgets('Triggers a Hero inside a ShellRoute',
        (WidgetTester tester) async {
      final UniqueKey heroKey = UniqueKey();

      final List<RouteBase> routes = <RouteBase>[
        ShellRoute(
          builder: (BuildContext context, GoRouterState state, Widget child) {
            return child;
          },
          routes: <GoRoute>[
            GoRoute(
                path: '/a',
                builder: (BuildContext context, _) {
                  return Hero(
                    tag: 'hero',
                    child: Container(),
                    flightShuttleBuilder: (_, __, ___, ____, _____) {
                      return Container(key: heroKey);
                    },
                  );
                }),
            GoRoute(
                path: '/b',
                builder: (BuildContext context, _) {
                  return Hero(
                    tag: 'hero',
                    child: Container(),
                  );
                }),
          ],
        )
      ];
      final GoRouter router =
          await createRouter(routes, tester, initialLocation: '/a');
      await tester.pumpAndSettle();

      // check that flightShuttleBuilder widget is not yet present
      expect(find.byKey(heroKey), findsNothing);

      // start navigation
      await tester.pump();
      router.go('/b');
      await tester.pump(const Duration(milliseconds: 30));
      expect(find.byKey(heroKey), isOnstage);

      // Waits for the animation finishes.
      await tester.pumpAndSettle();
      expect(find.byKey(heroKey), findsNothing);
    });
```

Comment (flodaniel) [reply packages/go_router/test/builder_test.dart:276]: @chunhtai @angjelkom I have fixed the test (something about when to `pump` was the issue. Can someone validate, commit and then merge this PR? 

```dart
testWidgets('Triggers a Hero inside a ShellRoute',
        (WidgetTester tester) async {
      final UniqueKey heroKey = UniqueKey();
      const String kHeroTag = 'hero';

      final List<RouteBase> routes = <RouteBase>[
        ShellRoute(
          builder: (BuildContext context, GoRouterState state, Widget child) {
            return child;
          },
          routes: <GoRoute>[
            GoRoute(
                path: '/a',
                builder: (BuildContext context, _) {
                  return Hero(
                    tag: kHeroTag,
                    child: Container(),
                    flightShuttleBuilder: (_, __, ___, ____, _____) {
                      return Container(key: heroKey);
                    },
                  );
                }),
            GoRoute(
                path: '/b',
                builder: (BuildContext context, _) {
                  return Hero(
                    tag: kHeroTag,
                    child: Container(),
                  );
                }),
          ],
        )
      ];
      final GoRouter router =
          await createRouter(routes, tester, initialLocation: '/a');

      // check that flightShuttleBuilder widget is not yet present
      expect(find.byKey(heroKey), findsNothing);

      // start navigation
      router.go('/b');
      await tester.pump();
      await tester.pump(const Duration(milliseconds: 10));
      // check that flightShuttleBuilder widget is visible
      expect(find.byKey(heroKey), isOnstage);
      // // Waits for the animation finishes.
      await tester.pumpAndSettle();
      expect(find.byKey(heroKey), findsNothing);
    });
```

**Comment by chunhtai** [comment]
Hi @angjelkom, it looks like @flodaniel has provided a test, can you commit to this pr so that i can merge it?

**Comment by angjelkom** [comment]
@flodaniel Thanks a lot for the test! @chunhtai merged it, but at the moment submit-queue check is failing for some other reason.

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by lublot** [comment]
Hello everyone! Is there any prediction of when it will be merged?

**Pull request merged**
